### ISSUE TITLE
More than one Excel category

### ISSUE THREAD

**Issue by mnar53** [sdllc/Basic-Excel-R-Toolkit#9]
It is ok to have all function prefixed with "R.". 

However, having all R functions grouped in one category can become rapidly impractical (too many).

It would be nice to have the capability to define more than one category to which assign R functions. Maybe just based on the file in which functions are defined (e.g., "Special Functions in R.R")

**Comment by duncanwerner** [comment]
Files probably not, but environments maybe.  As it stands you can use export functions from a named environment instead of .GlobalEnv, so it would just require extending that to multiple environments.

**Comment by mnar53** [comment]
If you mean              unhidden environment ==  excel category

this seems a viable solution.

**Comment by mnar53** [comment]
I think the newly introduced attribute could work fine for introducing multiple excel categories in BERT.
The best would be to have assignment of category AND function prefix, something like

attr(cholesky, "category") <-  list ("Linear Algebra", "linalg");

The processing done by RegisterAddinFunctions() at startup should produce an Excel category called "Linear Algebra", and in it the function linalg.cholesky. If function prefix absent, it defaults to category without space. If the attribute category absent, just work as now.

An imprecision in the docs: the Excel C-API is row-major, but in vba matrices are column-major.

**Comment by duncanwerner** [comment]
I don't want to remove the R. prefix, because it's too easy to abuse.  I definitely don't want to support functions without any prefix, because they can overwrite built-in functions.  Allowing custom prefixes is less of a problem, but a lot of add-ins use prefixing so this would allow overwriting of other libraries.

Given that, I think using an attribute for a prefix is unnecessary because you can just name the function using your desired prefix (we do support prefixing for importing packages, because then you don't have control over the function names).

The implementation in 1.08 uses that syntax but with just a string for the category:
```
attr( Cholesky, "category" ) <- "Linear Algebra";
```

and also in the package importing utility (with prefix):
```
BERT$UsePackage( "gsl", "gsl", category="GSL-BESSEL", pattern="bessel" );
```

**Comment by duncanwerner** [comment]
Regarding rows and columns: I probably shouldn't have phrased it that way, it's confusing.  VBA isn't really column-major or row-major, it only supports one-dimensional arrays (SAFEARRAY types).  In practice, you're correct that it will be column major if it has two dimensions, because the outer array reflects the rows (although for the same reason, if you have a one-dimensional array it's actually row major).

**Comment by mnar53** [comment]
I agree that a function prefix is necessary, but maybe one is enough. Let me explain: with the current use, 

USEPKG ('pracma', 'specfcn', pattern='erf')  -> 
                                                      Exported R Functions  ->  R.specfcn.erf
with the proposed solution

USEPKG ('pracma', 'specfcn', category='Special Functions',pattern='erf')  -> 

                                                      Special Functions  ->  R.specfcn.erf

I was just arguing that the R, in this case, becomes redundant. Not a big deal, after all if it becomes possible to define categories then i'd prefer to make explicit the originating package:

USEPKG ('pracma', 'pracma', category='Special Functions',pattern='erf')  -> 

                                                      Special Functions  ->  R.pracma.erf
or maybe both:

USEPKG ('pracma', 'specfcn.pracma', category='Special Functions',pattern='erf')  -> 

                                                      Special Functions  ->  R.specfcn.pracma.erf

In these cases too i feel the R. is redundant. As i said in the previous post, i would leave it as default, together with "Exported R Functions", if a category is NOT assigned. That means, without breaking
current usage. Given this, I'm eager to see published the 1.08 version.

**Comment by duncanwerner** [comment]
That's what it does now (what it's supposed to do -- if it's not doing that, then let me know).  I am just not ready to remove the R. prefix, even if you use your own prefix.  I agree it's redundant, but I'm being cautious.  

You can try the 1.08 release now, it's on the [releases page][1].  I am still testing the PATH change (see release notes) so it's not mainline yet, but I don't think it will cause significant problems.

[1]: https://github.com/sdllc/Basic-Excel-R-Toolkit/releases

**Comment by mnar53** [comment]
Got what you mean! I'd leave out the use of row-major / column-major, which are related to how bidimensional data is organized in the linear computer memory, and just state:

"In exposing a uni-dimensional vector to the Excel sheet, for example the output of a seq (linspace) function, one has to decide if a vector is by default a "row" or a "column" vector. Following the common usage of linear algebra, e.g. A x = b, bert interface defaults to COLUMN vector. This is the opposite of vba.

For the mapping of sheet data there is no confusion, beacause everything is mapped to a matrix, even
data in a row or data in a column"

**Comment by mnar53** [comment]
I've tried 1.08. I'm here reporting my experience:

I had some hard time because the system is a little bit "unforgivable": if there is an error it just does not execute the rest of the script. In my case I was filling the available slots and the USEPKG ..  where not reached. I've recompiled the code with more slots (by the way, better to stick to the Visual Studio compiler, for some reasons if compiled with Intel the xll does not work). 
Then I was not able to use BERT$USEPKG, probably my fault, and ended up using this version:

`USEPKG <- function( pkg, prefix, category, ... ){
	require (pkg,character.only=TRUE)
	BERT$UseEnvironment( paste0( "package:", pkg ), prefix, category, ... );
}`

It works. The I tried the attr, and even that works fine. However, consider this excerpt:

`CATEGORY <- 'Math'`
` Sqrt  <- function (x)  sqrt(x) `
`attr(Sqrt, "category" ) <- CATEGORY`
` Abs   <- function (x)  abs(x) `
`attr(Abs, "category" ) <- CATEGORY`

In R, for what I understand, there are no decorators as in Python (which allow to couple function definition and attribute). But definitively the code is a little bit redundant. To alleviate it I'm looking for something like

`mkattrs( list(Sqrt,Abs), "Math")`

But unfortunately i'm not able to code it correclty!
Anyway: good job!

**Comment by duncanwerner** [comment]
I made a small change to BERT$UsePackage (after 1.08) which may resolve your problem.  I would skip it for now.  There are no decorators in R, that would have been ideal but it's not available.  Attributes seemed like  the next-best solution.

One way to simplify might be to use an environment to hold your wrapper functions, then use UseEnvironment with a category:

```
MathFuncs <- new.env();
MathFuncs$SQRT <- function(x) sqrt(x)
MathFuncs$ABS <- function(x) abs(x)
BERT$UseEnvironment( MathFuncs, category="Math" )
```
if you're just wrapping functions, without making any changes to parameters, you can use the functions directly:
```
MathFuncs$SQRT <- sqrt
MathFuncs$ABS <- abs
```
and in fact if all you're doing is wrapping these functions, you could probably automate it.
